File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> preorderElements;
        
        stack<TreeNode*> helperStack;
        TreeNode* curNode = root;
        
        while(curNode != nullptr || !helperStack.empty()) {
            if(curNode == nullptr) {
                curNode = helperStack.top();
                helperStack.pop();
            }
            preorderElements.push_back(curNode->val);
            if(curNode->right) helperStack.push(curNode->right);
            
            curNode = curNode->left;
        }
        
        return preorderElements;
    }
};

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        Stack<TreeNode> st = new Stack<>();
        List<Integer> ans = new ArrayList<>();
        
        if(root == null) return ans;
        
        st.push(root);
        while(!st.isEmpty()) {
            TreeNode node = st.pop();
            ans.add(node.val);
            
            if(node.right != null) {
                st.push(node.right);
            }
            if(node.left != null) {
                st.push(node.left);
            }
            
        }
        
        return ans;
    }
}